storage/norcow_blockwise.h

```diff
@@ -0,0 +1,333 @@
+                            const uint16_t len) {
+  const void *ptr = NULL;
+  uint16_t allocated_len = 0;
+  if (sectrue != find_item(norcow_write_sector, key, &ptr, &allocated_len)) {
+    return secfalse;
+  }
+
+  if (allocated_len <= NORCOW_SMALL_ITEM_SIZE) {
+    // small items are not updated in place
+    return secfalse;
+  }
+
+  uint32_t sector_offset =
+      (const uint8_t *)ptr -
+      (const uint8_t *)norcow_ptr(norcow_write_sector, 0, NORCOW_SECTOR_SIZE);
+
+  const flash_area_t *area = &STORAGE_AREAS[norcow_write_sector];
+
+  if (norcow_write_buffer_key != key && norcow_write_buffer_key != -1) {
+    // some other update bytes is in process, abort
+    return secfalse;
+  }
+
+  if (norcow_write_buffer_key == -1) {
+    memset(norcow_write_buffer, 0xFF, sizeof(norcow_write_buffer));
+    norcow_write_buffer_key = key;
+    norcow_write_buffer_filled = 0;
+    norcow_write_buffer_flashed = 0;
+  }
+
+  if (norcow_write_buffer_flashed + norcow_write_buffer_filled + len >
+      allocated_len) {
+    return secfalse;
+  }
+
+  uint16_t tmp_len = len;
+  uint16_t flash_offset = sector_offset + norcow_write_buffer_flashed;
+
+  ensure(flash_unlock_write(), NULL);
+  while (tmp_len > 0) {
+    uint16_t buffer_space = FLASH_BLOCK_SIZE - norcow_write_buffer_filled;
+    uint16_t data_to_copy = (tmp_len > buffer_space ? buffer_space : tmp_len);
+    memcpy(&((uint8_t *)norcow_write_buffer)[norcow_write_buffer_filled], data,
+           data_to_copy);
+    data += data_to_copy;
+    norcow_write_buffer_filled += data_to_copy;
+    tmp_len -= data_to_copy;
+
+    bool all_data_received = (norcow_write_buffer_filled +
+                              norcow_write_buffer_flashed) == allocated_len;
+    bool block_full = norcow_write_buffer_filled == FLASH_BLOCK_SIZE;
+
+    if (block_full || all_data_received) {
+      if (!block_full) {
+        // all data has been received, add valid flag to last block
+        ((uint8_t *)norcow_write_buffer)[norcow_write_buffer_filled] =
+            NORCOW_VALID_FLAG;
+      }
+
+      ensure(flash_area_write_block(area, flash_offset, norcow_write_buffer),
+             NULL);
+      flash_offset += FLASH_BLOCK_SIZE;
+
+      if (block_full && all_data_received) {
+        // last block of data couldn't fit the valid flag, write it in next
+        // block
+        memset(norcow_write_buffer, 0xFF, sizeof(norcow_write_buffer));
+        ((uint8_t *)norcow_write_buffer)[0] = NORCOW_VALID_FLAG;
+        ensure(flash_area_write_block(area, flash_offset, norcow_write_buffer),
+               NULL);
+        flash_offset += FLASH_BLOCK_SIZE;
+      }
+
+      norcow_write_buffer_filled = 0;
+      norcow_write_buffer_flashed += FLASH_BLOCK_SIZE;
+      memset(norcow_write_buffer, 0xFF, sizeof(norcow_write_buffer));
+
+      if (all_data_received) {
+        norcow_write_buffer_key = -1;
+        norcow_write_buffer_flashed = 0;
+      }
+    }
+  }
```

Comment:
This looks OK, I think. Perhaps there is a cleaner way to write it, but I don't have time to think about it at the moment. We should test this code very thoroughly in the unit tests by creating and reading encrypted entries of all lengths between 0 to 132 bytes (so that up to two cipher blocks + another quadword are written).

Reply:
added the tests for the suggested lengths (for encrypted and unencrypted items too). I also fixed the python norcow_qr tests

https://github.com/trezor/trezor-firmware/commit/f5502791fdf64a5fa8a759cc6291738acb0294c8
https://github.com/trezor/trezor-firmware/commit/dfece8f91acdcac177bd901af7ff159a490292ea

Reply:
@andrewkozlik Can we resolve this conversation?